jgrapht-core/src/main/java/org/jgrapht/alg/tour/HamiltonianCycleAlgorithmBase.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * (C) Copyright 2019-2019, by Peter Harman and Contributors.
+ *
+ * JGraphT : a free Java graph-theory library
+ *
+ * See the CONTRIBUTORS.md file distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0, or the
+ * GNU Lesser General Public License v2.1 or later
+ * which is available at
+ * http://www.gnu.org/licenses/old-licenses/lgpl-2.1-standalone.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR LGPL-2.1-or-later
+ */
+package org.jgrapht.alg.tour;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import org.jgrapht.Graph;
+import org.jgrapht.GraphPath;
+import org.jgrapht.GraphTests;
+import org.jgrapht.alg.interfaces.HamiltonianCycleAlgorithm;
+import org.jgrapht.graph.GraphWalk;
+
+/**
+ * Base class for TSP solver algorithms.
+ *
+ * <p>
+ * This class provides implementations of utilities for TSP solver classes.
+ * </p>
+ *
+ * @param <V> the graph vertex type
+ * @param <E> the graph edge type
+ *
+ * @author Peter Harman
+ */
+public abstract class HamiltonianCycleAlgorithmBase<V, E> implements HamiltonianCycleAlgorithm<V, E> {
```

Comment:
I definitely agree with you. When looking at the preexisting algorithms there is potential to avoid code duplications by using checkGraph() and listToTour().

Additionally I thought splitting the getTour() implementations into two methods, first an initialization for a certain graph-instance (for algorithms that have a state like `TwoOptHeuristicTSP`) and second the actual tour computation for the initialized graph. With that approach repetitive tour computations to get the best of multiple tries could be handled unified by the base-class.
Also this would enable a generic `CombinedHeuristicTSP` where one can specify an initial-`HamiltonianCycleAlgorithm` and an afterwards improving `HamiltonianCycleImprovementAlgorithm` and this also for multiple tries without reinitializing the same graph again and again.

This would also improve the separation of the actual algorithm and accessory parts like input checks, repetitions to hope for better results and path computations. All these things could be handled in the base class uniformly. 
If the methods for initialization and the actual path computation are `protected` it is still not possible to break the algorithms from outside. Also the interface-API remains unchanged.

But I think this PR should be completed first.
What do you think about an Issue where we could discuss follow up improvements of the TSP package? I can open one.